### ISSUE TITLE
코드 리뷰 1

### ISSUE THREAD

**Issue by sseung0703** [seungwon1/batch_rl#8]
# 최적화 관련 이슈
개인적으로 더 보기 좋거나 효율적으로 짜는 방법이 있다고 생각하는 부분들 위주로 적었습니다.
실제로 시간 측정을 해보지 않은 부분이 있어서 약간 뇌피셜이 있고 손 코딩인지라 오류가 있을 수 있습니다. 하나의 의견 정도로 생각하시면 될 것 같습니다 😄 

### 73 lines
    l_vec = tf.transpose(tf.expand_dims(l, -1) * tf.ones([self.mini_batch, self.num_heads, self.num_heads]), perm = [0, 2, 1]) # (32, 51, 51)
위 코드는 아래와 같은 결과가 나오기 때문에 수정하는 것이 좋을 것 같습니다.

    l_vec = tf.expand_dims(l, 1) * tf.ones([self.mini_batch, self.num_heads, self.num_heads])
    l_vec = tf.tile(tf.expand_dims(l, 1), [1, self.num_heads, 1])

### 80 lines
    prob_ml = tf.reduce_sum(qtarget_vec_ml * tf.cast(tf.equal(l_vec, idx_arr), tf.float32), axis = 2) # (32, 51)
*tf.cast(tf.equal)는 직관적이지만 무의미한 곱연산보다는 아래와 같이 tf.where를 사용하는 것이 좋을 것 같습니다.

    prob_ml = tf.reduce_sum(tf.where(tf.equal(l_vec, idx_arr), qtarget_vec_ml, tf.zeros_like(qtarget_vec_ml), axis = 2) # (32, 51)

### 86-87 lines
    self.sess.run(tf.global_variables_initializer())    
    self.sess.run([tf.assign(t, o) for t, o in zip(self.var_target, self.var_online)])

global variables를 정해진 intializer로 initialize한 다음 원하는 parameter로 다시 initialize하는 것이기 때문에 효율적이지 않습니다. 아래와 같이 수정하면 한 번에 원하는 parameter로 initialize할 수 있습니다.

    for t, o in zip(self.var_target, self.var_online):
        t._initial_value = o
        t._initializer_op = tf.assign(t._variable, t._initial_value, name= v.name[:-2]+'/Assign').op
    self.sess.run(tf.global_variables_initializer())

### 128-129 lines
    # Reset target_variable in every target_reset interval
    if step_count % self.FLAGS.target_reset == 0:
        self.sess.run( [tf.assign(t, o) for t, o in zip(self.var_target, self.var_online)])
알고리즘을 전부 파악한 것이 아니라 정확하지 않지만 self.var_online이 변하지 않는 상황이고 위에서 제가 말씀드린 방식대로 initializer를 변경하였다면 아래와 같이 바꿀 수 있습니다.
    self.sess.run( [t.initializer for t in self.var_target] )

### 57 lines
    rgb2y = 0.299*images[...,0] + 0.587*images[...,1] + 0.114*images[...,2] # compute luminance from rgb

rgb를 gray로 바꾸기 위해 slicing --> multiply --> addition을 해주셨습니다. 이는 아래와 같이 dot으로 한 번에 처리할 수 있습니다.

    rgb2y = images.dot([0.299, 0.587, 0.114])
